MOLT Fetch GA fast-follow issues (#18573)
cockroachdb/docs: src/current/v23.2/molt-fetch.md

| Linux            | [Download](https://molt.cockroachdb.com/molt/cli/molt-latest.linux-amd64.tgz)   | [Download](https://molt.cockroachdb.com/molt/cli/molt-latest.linux-arm64.tgz)   |
| Mac              | [Download](https://molt.cockroachdb.com/molt/cli/molt-latest.darwin-amd64.tgz)  | [Download](https://molt.cockroachdb.com/molt/cli/molt-latest.darwin-arm64.tgz)  |

> For previous binaries, see the [MOLT version manifest](https://molt.cockroachdb.com/molt/cli/versions.html). 

Reviewer: ```suggestion
For previous binaries, refer to the [MOLT version manifest](https://molt.cockroachdb.com/molt/cli/versions.html). 
```

{{site.data.alerts.callout_info}}

> Red Hat Enterprise Linux (RHEL) 9 and above are supported.

Reviewer: ```suggestion
MOLT Fetch is supported on Red Hat Enterprise Linux (RHEL) 9 and above.
```

{{site.data.alerts.end}}

## Setup

Complete the following items before using MOLT Fetch:

- Follow the recommendations in [Best practices](#best-practices) and [Security recommendations](#security-recommendations).

> - Ensure that the source and target schemas are identical, unless you enable automatic schema creation with the [`'drop-on-target-and-recreate'`](#target-table-handling) option. If you are manually creating the target schema, review the behaviors in [Mismatch handling](#mismatch-handling).

Reviewer: ```suggestion
- Ensure that the source and target schemas are identical, unless you enable automatic schema creation with the [`'drop-on-target-and-recreate'`](#target-table-handling) option. If you are creating the target schema manually, review the behaviors in [Mismatch handling](#mismatch-handling).
```

- If a MySQL database is set as a [source](#source-and-target-databases), the [`--table-concurrency`](#global-flags) and [`--export-concurrency`](#global-flags) flags **cannot** be set above `1`. If these values are changed, MOLT Fetch returns an error. This guarantees consistency when moving data from MySQL, due to MySQL limitations. MySQL data is migrated to CockroachDB one table and shard at a time, using [`WITH CONSISTENT SNAPSHOT`](https://dev.mysql.com/doc/refman/8.0/en/commit.html) transactions.

> - To prevent memory outages during data export of tables with large rows, estimate the amount of memory used to export a table: `--row-batch-size * --export-concurrency * average size of the table rows`. If you are exporting more than one table at a time (i.e., [`--table-concurrency`](#global-flags) is set higher than `1`), add the estimated memory usage for the tables with the largest row sizes. Ensure that you have sufficient memory to run `molt fetch`, and adjust `--row-batch-size` accordingly.

Reviewer: ```suggestion
- To prevent memory outages during data export of tables with large rows, estimate the amount of memory used to export a table: 

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    --row-batch-size * --export-concurrency * average size of the table rows
    ~~~
    
    If you are exporting more than one table at a time (i.e., [`--table-concurrency`](#global-flags) is set higher than `1`), add the estimated memory usage for the tables with the largest row sizes. Ensure that you have sufficient memory to run `molt fetch`, and adjust `--row-batch-size` accordingly.
```

I think the formula is easier to read when it's in a block. I didn't verify the formatting locally.

> - If a table in the source database is much larger than the other tables, [filter and export the largest table](#schema-and-table-selection) in its own `molt fetch` task. Then export the remaining tables in another task.

Reviewer 2: I think it's worth noting that this process should be repeated for all outsized tables

| `--pprof-listen-addr`                         | Address of the pprof endpoint.<br><br>**Default:** `'127.0.0.1:3031'`                                                                                                                                                                                                                                                                                                                                                          |
| `--replicator-flags`                          | If continuous [replication](#replication) is enabled with `--ongoing-replication`, specify Replicator flags ([PostgreSQL](https://github.com/cockroachdb/replicator/wiki/PGLogical#postgresql-logical-replication) or [MySQL](https://github.com/cockroachdb/replicator/wiki/MYLogical#mysqlmariadb-replication)) to override.                                                                                                 |

> | `--row-batch-size`                            | Number of rows per shard to export at a time. See [Best practices](#best-practices).<br><br>**Default:** `100000`                                                                                                                                                                                                                                                                                                              |

Reviewer: ```suggestion
| `--row-batch-size`                            | Number of rows per shard to export at a time. Refer to [Best practices](#best-practices).<br><br>**Default:** `100000`                                                                                                                                                                                                                                                                                                              |
```

### Fetch mode

MOLT Fetch can use either [`IMPORT INTO`]({% link {{ page.version.version }}/import-into.md %}) or [`COPY FROM`]({% link {{ page.version.version }}/copy-from.md %}) to load data into CockroachDB.

By default, MOLT Fetch uses `IMPORT INTO`:

> - `IMPORT INTO` mode achieves the highest throughput, but [requires taking the tables **offline**]({% link {{ page.version.version }}/import-into.md %}#considerations) to achieve its import speed. Tables are taken back online once an [import job]({% link {{ page.version.version }}/import-into.md %}#view-and-control-import-jobs) completes successfully. See [Best practices](#best-practices).

Reviewer: ```suggestion
- `IMPORT INTO` mode achieves the highest throughput, but [takes the tables **offline**]({% link {{ page.version.version }}/import-into.md %}#considerations) to achieve its import speed. Tables are taken back online once an [import job]({% link {{ page.version.version }}/import-into.md %}#view-and-control-import-jobs) completes successfully. Refer to [Best practices](#best-practices).
```

When using the `'drop-on-target-and-recreate'` option, MOLT Fetch creates a new CockroachDB table to load the source data if one does not already exist. To guide the automatic schema creation, you can [explicitly map source types to CockroachDB types](#type-mapping).

#### Mismatch handling

> If either [`'none'`](#target-table-handling) or [`'truncate-if-exists'`](#target-table-handling) is set, `molt fetch` loads data into the existing tables on the target CockroachDB database. If the target schema mismatches the source schema, `molt fetch` will exit early in certain cases, and will need to be re-run from the beginning.

Reviewer: ```suggestion
If either [`'none'`](#target-table-handling) or [`'truncate-if-exists'`](#target-table-handling) is set, `molt fetch` loads data into the existing tables on the target CockroachDB database. If the target schema mismatches the source schema, `molt fetch` will exit early in [certain cases](#exit-early), and will need to be re-run from the beginning.
```

{{site.data.alerts.callout_info}}

> This does not apply when [`'drop-on-target-and-recreate'`](#target-table-handling) is specified, since this mode automatically creates a compatible CockroachDB schema.

Reviewer: ```suggestion
When [`'drop-on-target-and-recreate'`](#target-table-handling) is specified, a compatible CockroachDB schema is created automatically.
```

{{site.data.alerts.end}}

> `molt fetch` exits early in the following cases, and will output a log with a corresponding `mismatch_tag` and `failable_mismatch` set to `true`:

Reviewer: ```suggestion
<a id="exit-early"></a>`molt fetch` exits early in the following cases, and will output a log with a corresponding `mismatch_tag` and `failable_mismatch` set to `true`:
```
